File: app/src/main/java/com/zuk0/gaijinsmash/riderz/ui/adapter/estimate/EstimateRecyclerAdapter.java
package com.zuk0.gaijinsmash.riderz.ui.adapter.estimate;

import android.os.CountDownTimer;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.zuk0.gaijinsmash.riderz.R;
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Estimate;
import com.zuk0.gaijinsmash.riderz.ui.fragment.home.HomeViewModel;
import com.zuk0.gaijinsmash.riderz.utils.BartRoutesUtils;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class EstimateRecyclerAdapter extends RecyclerView.Adapter<EstimateRecyclerAdapter.ViewHolder>{

    static class ViewHolder extends RecyclerView.ViewHolder {
        private TextView origin;
        private TextView destination;
        private TextView minutes;
        private TextView line;

        ViewHolder(View view) {
            super(view);
            origin = view.findViewById(R.id.etd_originTv);
            destination = view.findViewById(R.id.etd_destinationTv);
            minutes = view.findViewById(R.id.etd_minutesTv);
            line = view.findViewById(R.id.etd_colored_line);
        }
    }

    private List<Estimate> mEstimateList;

    public EstimateRecyclerAdapter(List<Estimate> estimateList) { mEstimateList = estimateList; }

    public List<CountDownTimer> timers = new ArrayList<>();


    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_row_estimate, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Estimate estimate = mEstimateList.get(position);
        holder.origin.setText(estimate.getOrigin());
        holder.destination.setText(estimate.getDestination());

        String minutes = estimate.getMinutes();
        if(!minutes.equalsIgnoreCase("Leaving")) {
            timers.add(HomeViewModel.beginTimer(holder.minutes, Integer.valueOf(minutes)));
        } else {
            holder.minutes.setText(R.string.leaving);
        }
        BartRoutesUtils.setLineBarByColor(holder.line.getContext(), estimate.getColor(), holder.line);
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        Log.i("recycler", "onDetachedFromRecyclerView");
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull ViewHolder holder) {
        super.onViewDetachedFromWindow(holder);
        timers.clear();
        Log.d("onDetachedFromView", "timers cleared");
    }

    @Override
    public int getItemCount() {
        return mEstimateList.size();
    }
}

File: app/src/main/java/com/zuk0/gaijinsmash/riderz/ui/activity/main/fragment/home/adapter/etd/EstimateRecyclerAdapter.kt
package com.zuk0.gaijinsmash.riderz.ui.activity.main.fragment.home.adapter.etd

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Adapter
import com.zuk0.gaijinsmash.riderz.R.string
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Estimate
import com.zuk0.gaijinsmash.riderz.databinding.ListColEstimateBinding
import com.zuk0.gaijinsmash.riderz.ui.activity.main.fragment.home.HomeViewModel
import com.zuk0.gaijinsmash.riderz.utils.BartRoutesUtils
import com.zuk0.gaijinsmash.riderz.utils.TrainTimer.Companion.TIMER_FINISHED

class
EstimateRecyclerAdapter(val estimateList: MutableList<Estimate>, val viewModel: HomeViewModel) : Adapter<EstimateRecyclerAdapter.ViewHolder>() {

    class ViewHolder(val binding: ListColEstimateBinding) : RecyclerView.ViewHolder(binding.root) {
    }

    //todo one master timer for the longest timeToWait
    //each view holder observes time of the master timer.
    //calculate difference every minute.
    //every 60 seconds, push a new value to observers.

    //private val timers: MutableList<CountDownTimer> = ArrayList()
    private lateinit var binding: ListColEstimateBinding

    override fun getItemId(position: Int): Long {
        return super.getItemId(position)
    }

    override fun getItemViewType(position: Int): Int {
        return ETD_VIEW_TYPE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        binding = ListColEstimateBinding.inflate(LayoutInflater.from(parent.context))
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val estimate = estimateList[position]

        //set car length of train
        val carLength = estimate.length.toString()
        holder.binding.estimateResultCarTv.text = carLength

        //set time remaining
        holder.binding.estimateResultMinutesTv.text = estimate.minutes

        val minutes: String? = estimate.minutes
        if (!minutes.equals("Leaving", ignoreCase = true)) {

            viewModel.timePassedLiveData.observeForever { amountOfTimePassed ->
                if(amountOfTimePassed == TIMER_FINISHED || amountOfTimePassed == 0L) {
                    holder.binding.estimateResultMinutesTv.text = "0"
                    //todo close view

                } else {
                    val remainingMinutes = holder.binding.estimateResultMinutesTv.text.toString().toInt()
                    val updatedTime = remainingMinutes - 1
                    val min = holder.binding.root.context.resources.getString(string.minutes)
                    val msg = "$updatedTime $min"
                    holder.binding.estimateResultMinutesTv.text = msg
                }
            }
        } else {
            holder.binding.estimateResultMinutesTv.text = "0"
            //todo close view
        }

        BartRoutesUtils.setLineBarByColor(holder.binding.background.context, estimate.color, holder.binding.estimateResultTrainIcon)
    }


    fun removeEstimate(position: Int) {
        val newList: MutableList<Estimate> = mutableListOf<Estimate>()
        newList.addAll(estimateList)
        newList.removeAt(position)
        update(newList)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        Log.i("recycler", "onDetachedFromRecyclerView")
    }

    override fun onViewDetachedFromWindow(holder: ViewHolder) {
        super.onViewDetachedFromWindow(holder)
        Log.d("onDetachedFromView", "timers cleared")
    }

    override fun getItemCount(): Int {
        return estimateList.size
    }

    fun update(newData: List<Estimate>) {
        val diffResult = DiffUtil.calculateDiff(EstimateDiffCallback(estimateList, newData))
        estimateList.clear()
        estimateList.addAll(newData)
        diffResult.dispatchUpdatesTo(this)
    }

    companion object {
        private const val TITLE_VIEW_TYPE = 1000
        private const val ETD_VIEW_TYPE = 2000
        private const val FAV_VIEW_TYPE = 2000
    }

}